### DISCUSSION TITLE
My first dictionary "book" using SILE as an XML processor for a TEI subset

### DISCUSSION THREAD

**Discussion by Omikhleia** [sile-typesetter/sile#1252] [Show and tell]
I have been the main editor of an XML-based dictionary about one of Tolkien's invented languages for some years. The source XML dictionary is encoded in a specialized subset of the [TEI P4 tagset](https://www.tei-c.org/Vault/P4/doc/html/DI.html). I fancied making a PDF out of it, using SILE. Would be fun, I said to myself, if SILE could help me where my previous attempts with XSLT, XSL-FO and FOP failed short (too many layers + too weird language = too many headaches)

So these past days, I wrote:

- A **teibook** class, with the bare minimum support for such dictionaries: running headers ("first-entry - folio - last-entry"), appropriate page masters (one column for title page and front matter, then two columns for dictionary entries) and related sectioning hooks, plus loading all needed packages.
  -  That's my first try at page masters, it's real interesting! I hope to refactor that part in a nicer way eventually.
  -  Regarding headers, I'm glad I had worked on the TOC recently, so I knew "infonodes" were what I had to look for :)
- A **teidict** package that does the actual work processing the XML content (tags etc.)
  - I am happy to have worked on the "url" packaged recently, as I then knew a similar "inputfilter" would help me breaking those long unwieldy bibliographic references the dictionary contains
  - Lots of AST magic and tricks... (More on that below) - I do hope the AST is somewhat stable :roll_eyes:    
- And of course a small preamble, nothing fancy here.

The pain point is that such a dictionary is a heavily "semantic" structured mark-up (i.e. a "lexical view", encoding structure information such as part-of-speech etc. without much concern for its exact textual representation in print form), much more than a "presentational" mark-up. Some XML nodes may contain many things we need to ignore (such as spaces, mostly) or supplement
(such as punctuation, parentheses, numbering - and again, proper spaces where needed). Without XPath to check siblings, ascendants or descendants, it may become somewhat hard to get a nice automated output (and even _with_ XPath, it's quite messy anyway). I am not fully satisfied with the current solution (and some _ad hoc_ tricks that might not be easy to generalize to _any_ XML TEI dictionary...), so I'll need to step back and re-think it eventually.

But _heh_, to make a long story short, I am glad to share my [current result](https://omikhleia.github.io/sindict/sile-sd-en.pdf), not perfect, but decently-looking (IMHO).

And the supporting class and packages are in my [other repository](https://github.com/Omikhleia/omikhleia-sile-packages)

**Comment by OlivierNicole** [comment]
i haven't taken the time to look at the XML processing, but the result looks nice!
